### PR TITLE
Initial addition of staBal routes.

### PR DESCRIPTION
**Pull request by johngrantuk**

### CODE REVIEW

File: src/routeProposal/filtering.ts [hunk 1]
```diff
@@ -39,8 +41,9 @@ export function filterPoolsOfInterest(
     tokenIn: string,
     tokenOut: string,
     maxPools: number,
+    chainId: number,
     currentBlockTimestamp = 0
-): [PoolDictionary, string[]] {
+): [PoolDictionary, string[], StablePool] {
```
Comment (TomAFrench) [inline src/routeProposal/filtering.ts:46]: I'm not a fan of functions which return many different things. Can we pull this StablePool out elsewhere?

**Comment by TomAFrench** [comment]
It's not really clear what's going on here, I think we need to add more documentation on what the flow of the algorithm is doing. It looks like we're going to be passing along incomplete paths to later stages of the algorithm.

File: test/staBalPaths.spec.ts [hunk 2]
```diff
@@ -0,0 +1,355 @@
+// TS_NODE_PROJECT='tsconfig.testing.json' npx mocha -r ts-node/register test/staBalPaths.spec.ts
```
Comment (TomAFrench) [inline test/staBalPaths.spec.ts:1]: Many of these tests are pretty much copy pastes from each other. Can you deduplicate these where you're just passing in different inputs to the same code?

This is something that we need to be doing generally with this codebase. The amount of duplication in the tests is huge.

File: src/routeProposal/filtering.ts [hunk 3]
```diff
@@ -33,15 +35,20 @@ The main purpose of this function is to:
...
 - parsePoolPairData for Direct pools
 - store token decimals for future use
+- We also create and return a pools dictionary that includes all pools which is useful for future pool filtering
 */
 export function filterPoolsOfInterest(
     allPools: SubgraphPoolBase[],
     tokenIn: string,
     tokenOut: string,
     maxPools: number,
+    chainId: number,
```
Comment (TomAFrench) [inline src/routeProposal/filtering.ts:45]: This is unused.

File: src/routeProposal/filtering.ts [hunk 4]
```diff
@@ -33,15 +35,20 @@ The main purpose of this function is to:
...
     currentBlockTimestamp = 0
-): [PoolDictionary, string[]] {
-    const poolsDictionary: PoolDictionary = {};
+): [PoolDictionary, string[], PoolDictionary] {
+    // This will include all pools
+    const poolsAllDictionary: PoolDictionary = {};
```
Comment (TomAFrench) [inline src/routeProposal/filtering.ts:49]: Personally I'd create this dictionary outside of this function and then pass it in to be filtered. We're overloading what the purpose of this function is by doing this.


My solution would look something like this:
```
const poolsDict = Object.fromEntries(pools
  .map((pool) => [pool.id, parseNewPool(pool, swapOptions.timestamp)])
  .filter(([,pool])=> pool !== undefined)
)
const [poolsFilteredDict, hopTokens] =
    filterPoolsOfInterest(
        poolsDict,
        tokenIn,
        tokenOut,
        swapOptions.maxPools,
    );
```

File: src/constants.ts [hunk 5]
```diff
@@ -25,6 +25,48 @@ export const VAULTADDR: { [chainId: number]: string } = {
     42161: '0xBA12222222228d8Ba445958a75a0704d566BF2C8',
 };
 
+// Currently only used for Polygon
+// ID of USDC Connecting Pool & USDC token address
+export const USDCCONNECTINGPOOL: {
+    [chainId: number]: { id: string; usdc: string };
+} = {
+    1: {
```
Comment (TomAFrench) [inline src/constants.ts:33]: Lets remove these keys if we're not using them

**Review by TomAFrench — CHANGES_REQUESTED** [review]
I've pushed a couple of changes to mostly rename some variables/comments and have better return values for some functions.

I think we're still overloading `filterPoolsOfInterest` however and the tests can be deduplicated. Happy with the general algo though.

File: test/staBalPaths.spec.ts [hunk 6]
```diff
@@ -0,0 +1,355 @@
...
+
+const maxPools = 4;
+const chainId = 99;
+
+describe(`staBalPaths.`, () => {
+    it(`should be no USDC connecting pool for mainnet`, () => {
+        const tokenIn = '0x0000000000085d4780B73119b644AE5ecd22b376';
+        const tokenOut = '0x9a71012b13ca4d3d0cdc72a177df3ef03b0e76a3'; // BAL
+
+        const testPools: any = cloneDeep(staBalPools.pools);
```
Comment (TomAFrench) [inline test/staBalPaths.spec.ts:25]: Why are these being converted to `any`?

File: test/staBalPaths.spec.ts (hunk 2 above)
Comment (johngrantuk) [reply test/staBalPaths.spec.ts:1]: Deduplicated and added more thorough Path testing.

File: src/routeProposal/filtering.ts (hunk 4 above)
Comment (johngrantuk) [reply src/routeProposal/filtering.ts]: Created a new function `parseToPoolsDict`. This also filters out pools with 0 balance.

**Review by TomAFrench — APPROVED** [review]
~~LGTM. @johngrantuk can you check you're happy with these final changes?~~

Ah, forgot you're improving some of the tests, nvm.

**Comment by johngrantuk** [comment]
@TomAFrench I've changed the previously highlighted tests to use all pools. Also updated where required to make things work after the removal of BigNumber.js. Would probably benefit from another quick check.

`constants.ts` needs updated with `USDCCONNECTINGPOOL` and `STABALADDR` information once those pools are deployed on Polygon so not sure if we merge before or after that?

**Comment by TomAFrench** [comment]
> I've changed the previously highlighted tests to use all pools. Also updated where required to make things work after the removal of BigNumber.js. Would probably benefit from another quick check.

Thanks, I'll give this a check over in a bit and then merge.

> constants.ts needs updated with USDCCONNECTINGPOOL and STABALADDR information once those pools are deployed on Polygon so not sure if we merge before or after that?

This isn't an issue. We just need to remove the entries for each networks and we'll fall straight through the new function so we can merge now.

**Pull request merged**